DMD.C

```diff
@@ -0,0 +1,489 @@
+                  int d_rank,
+                  int d_num_procs)
+{
+    Matrix* f_snapshots_minus = new Matrix(f_snapshots->numRows(),
+                                           f_snapshots->numColumns() - 1, f_snapshots->distributed());
+    Matrix* f_snapshots_plus = new Matrix(f_snapshots->numRows(),
+                                          f_snapshots->numColumns() - 1, f_snapshots->distributed());
+
+    // Break up snapshots into snapshots_minus and snapshots_plus
+    // snapshots_minus = all columns of snapshots except last
+    // snapshots_plus = all columns of snapshots except first
+    for (int i = 0; i < f_snapshots->numRows(); i++)
+    {
+        f_snapshots_minus->item(i, 0) = f_snapshots->item(i, 0);
+        for (int j = 1; j < f_snapshots->numColumns() - 1; j++)
+        {
+            f_snapshots_minus->item(i, j) = f_snapshots->item(i, j);
+            f_snapshots_plus->item(i, j - 1) = f_snapshots->item(i, j);
+        }
+        f_snapshots_plus->item(i, f_snapshots->numColumns() - 2) =
+            f_snapshots->item(i, f_snapshots->numColumns() - 1);
+    }
+
+    int *row_offset = new int[d_num_procs + 1];
+    row_offset[d_num_procs] = f_snapshots_minus->numDistributedRows();
+    row_offset[d_rank] = f_snapshots_minus->numRows();
+
+    CAROM_VERIFY(MPI_Allgather(MPI_IN_PLACE,
+                               1,
+                               MPI_INT,
+                               row_offset,
+                               1,
+                               MPI_INT,
+                               MPI_COMM_WORLD) == MPI_SUCCESS);
+    for (int i = d_num_procs - 1; i >= 0; i--) {
+        row_offset[i] = row_offset[i + 1] - row_offset[i];
+    }
+
+    CAROM_VERIFY(row_offset[0] == 0);
+
+    int d_blocksize = row_offset[d_num_procs] / d_num_procs;
+    if (row_offset[d_num_procs] % d_num_procs != 0) d_blocksize += 1;
+
+    SLPK_Matrix svd_input;
+
+    // Calculate svd of snapshots_minus
+    initialize_matrix(&svd_input, f_snapshots_minus->numColumns(),
+                      f_snapshots_minus->numDistributedRows(),
+                      1, d_num_procs, d_blocksize, d_blocksize);
+
+    for (int rank = 0; rank < d_num_procs; ++rank)
+    {
+        scatter_block(&svd_input, 1, row_offset[rank] + 1,
+                      f_snapshots_minus->getData(),
+                      f_snapshots_minus->numColumns(),
+                      row_offset[rank + 1] - row_offset[rank], rank);
+    }
+
+    std::unique_ptr<SVDManager> d_factorizer(new SVDManager);
+
+    // This block does the actual ScaLAPACK call to do the factorization.
+    svd_init(d_factorizer.get(), &svd_input);
+    d_factorizer->dov = 1;
+    factorize(d_factorizer.get());
+    free_matrix_data(&svd_input);
+
+    // Compute how many basis vectors we will actually use.
+    int num_singular_vectors = std::min(f_snapshots_minus->numColumns(), f_snapshots_minus->numDistributedRows());
+    if (d_energy_fraction != -1)
```

Comment:
This must work, but it still worries me a little to compare a double and int. Why not make it `-1.0`?